concrete-core/src/crypto/gsw/levels.rs

```diff
@@ -0,0 +1,247 @@
+use crate::crypto::lwe::LweCiphertext;
+use crate::math::decomposition::DecompositionLevel;
+use crate::math::tensor::{AsMutTensor, AsRefSlice, AsRefTensor, Tensor};
+use crate::tensor_traits;
+use concrete_commons::parameters::LweSize;
+#[cfg(feature = "multithread")]
+use rayon::prelude::*;
+
+/// A matrix containing a single level of gadget decomposition.
+pub struct GswLevelMatrix<Cont> {
+    tensor: Tensor<Cont>,
+    lwe_size: LweSize,
+    level: DecompositionLevel,
+}
+
+tensor_traits!(GswLevelMatrix);
+
+impl<Cont> GswLevelMatrix<Cont> {
+    /// Creates a GSW level matrix from an arbitrary container.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use concrete_commons::parameters::LweSize;
+    /// use concrete_core::crypto::gsw::GswLevelMatrix;
+    /// use concrete_core::math::decomposition::DecompositionLevel;
+    /// let level_matrix =
+    ///     GswLevelMatrix::from_container(vec![0 as u8; 7 * 7], LweSize(7), DecompositionLevel(1));
+    /// assert_eq!(level_matrix.lwe_size(), LweSize(7));
+    /// assert_eq!(level_matrix.decomposition_level(), DecompositionLevel(1));
+    /// ```
+    pub fn from_container(cont: Cont, lwe_size: LweSize, level: DecompositionLevel) -> Self
+    where
+        Cont: AsRefSlice,
+    {
+        let tensor = Tensor::from_container(cont);
+        GswLevelMatrix {
```

Comment:
Could we have a check here to verify that the lwe size divides the container length ? Maybe it would be nice to check that the result of this division is also lwe_size.

Reply:
The check has a bug. Here `level` is a `DecompositionLevel`, which represent the index of the level this matrix corresponds to in the decomposition. The check should be:
```
        ck_dim_eq!(tensor.as_slice().len() => lwe_size.0 * lwe_size.0);
```